### PR TITLE
Create a new PR for paging support in Android core

### PR DESCRIPTION
**Pull request by JianpingChen**
Add basic paging support. 

Design goals:
1. Support autorest.android generating Async client. Required classes: Page, AsyncPagedDataRetriever, AsyncPagedDataCollection
2. Support autorest.android generating Sync client. Required classes: Page, PagedDataCollection, PagedDataRetriever, PagedDataResponseCollection, PagedDataResponseRetriever

### CODE REVIEW

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java
Comment (bsiegel) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java:112]: We should be using `com.azure.android.core.http.Response<T>` as the type to represent a response with an associated strongly-typed response body object. Is there any reason we aren't doing that here?

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataRetriever.java
Comment (bsiegel) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataRetriever.java:50]: How is the pageId supposed to be communicated to the developer? Is it an opaque identifier (e.g. GUID)? I would expect `getPage(int pageNumber, Callback callback)` or something similar, otherwise the developer has to do book-keeping.

Alternatively, why not keep the current page number in the instance's local state, and have a `getNextPage(Callback callback)` method (and a simple getter, `getCurrentPageNumber()`or some such) instead?

Comment (bsiegel) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataRetriever.java:50]: I guess this comment isn't specific to this file, but just to the general design.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseCollection.java
Comment (bsiegel) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseCollection.java:19]: Is this the sync version of `AsyncPagedDataCollection`? If so, they should have similar names. E.g. `PagedDataCollection`.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseRetriever.java
Comment (bsiegel) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseRetriever.java:17]: Likewise if this is the sync version of `AsyncPagedDataRetriever`, let's make this `PagedDataRetriever`.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseRetriever.java
Comment (bsiegel) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseRetriever.java:13]:
```suggestion
 * Represents a synchronous paged data provider with fixed page size that returns both page and response
```

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java
Comment (JianpingChen) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java:112]: Sorry Brandon, This branch is messed up somehow. Please put comment in https://github.com/Azure/azure-sdk-for-android/pull/317 instead. I'll still read here but more people may have been directed there.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PagedDataCollection.java
Comment (bsiegel) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PagedDataCollection.java:11]:
```suggestion
 * Represents a collection of paged data with fixed page size where pages can be retrieved synchronously
```

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PagedDataCollection.java
Comment (bsiegel) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PagedDataCollection.java:15]: Not clear why there is a separate set of classes for paged data vs. paged data + response, but only for sync (for async there is only paged data + response).

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java
Comment (JianpingChen) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java:112]: My bad. I thought this was the old one. ;-)

The reason for PageAndResponse is this collection has an interface where getFirstPage and getPage both take a Callback<P> where both Page and okhttp3.Response are supposed to be used. Unless we change the callback signature, we'd have to keep both in order to hand back the okhttp3.Response.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataRetriever.java
Comment (JianpingChen) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataRetriever.java:50]: For the need of autorest.android, the String id is meant to be opaque. I even had a page iterator function in the collection trying to hide the pageId altogether but it turned out to be overly complicated (in thread synchronization). So we agreed to remove the iterator implementation. 

We looked at using this interface to implement androidx.arch.DataSource in my initial Draft PR. Having an opaque page id is more convenient than int id for that particular purpose. That's the reason we settled with an opaque string id.

c.f https://developer.android.com/reference/android/arch/paging/PageKeyedDataSource#loadBefore(android.arch.paging.PageKeyedDataSource.LoadParams%3CKey%3E,%20android.arch.paging.PageKeyedDataSource.LoadCallback%3CKey,%20Value%3E)

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseCollection.java
Comment (JianpingChen) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseCollection.java:19]: There is a PagedDataCollection in util.paging. We can rename AsyncPagedDataCollection to AsyncPagedDataResponseCollection

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseRetriever.java
Comment (JianpingChen) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseRetriever.java:17]: There is a PagedDataRetriever in util.paging which does not return Response.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PagedDataCollection.java
Comment (JianpingChen) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PagedDataCollection.java:15]: When autorest.android generates sync client, for each function in underlying service, e.g. service.getBlah(), sync client is supposed to have two versions of the same function in its interface, syncClient.getBlah(): Blah, and syncClient.getBlahWithResponse(): Response\<Blah\>. That's why we have two sets for sync client.

Comment (JianpingChen) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PagedDataCollection.java:15]: For async client, the Callback\<Blah\> can return both Blah and okhttp3.Response. Therefore there is no need to have two function signatures for each service.getBlah function.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataRetriever.java
Comment (anuchandy) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataRetriever.java:52]: Thinking a bit more about it: 

The paged collection = retrieve pages from service via network + caching of the pages.

Wondering, whether it makes sense to have the following signature:

```java
// group_1: calls go over network.
void getFirstPage(Callback<Page<T>> callback);
void getNextPage(Callback<Page<T>> callback);
```

```java
// group_2: an immutable map of cached pages.
Map<String, Page<T>> getCachedPages();
// an immutable list of items in the cached pages.
List<T> getCachedItems();
```

Only the APIs that belongs to the group_1 provides access to the Response object.

**Comment by bsiegel** [comment]
@JianpingChen was there anything else we needed to do to this PR post today's discussion?

File: sdk/core/azure-core/src/test/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollectionTest.java
Comment (vcolin7) [inline sdk/core/azure-core/src/test/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollectionTest.java:1]: We need to add the license header here.

File: sdk/core/azure-core/src/test/java/com/azure/android/core/http/responsepaging/PagedDataResponseCollectionTest.java
Comment (vcolin7) [inline sdk/core/azure-core/src/test/java/com/azure/android/core/http/responsepaging/PagedDataResponseCollectionTest.java:1]: We need to add the license header here.

File: sdk/core/azure-core/src/test/java/com/azure/android/core/util/paging/PagedDataCollectionTest.java
Comment (vcolin7) [inline sdk/core/azure-core/src/test/java/com/azure/android/core/util/paging/PagedDataCollectionTest.java:1]: We need to add the license header here.

**Comment by JianpingChen** [comment]
> @JianpingChen was there anything else we needed to do to this PR post today's discussion?

@bsiegel As far as I remember, no. The idea of making it Page<K, T> was rejected. Arguments around setters for nexpPage and previousPage seems to settle with "as is", unless I misunderstood anything. We can make it clear all collections are caching in JavaDoc if that is required.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java
Comment (bsiegel) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java:119]: Leaving as-is for this preview.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataRetriever.java
Comment (bsiegel) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataRetriever.java:52]: Leaving as-is for this preview.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseCollection.java
Comment (bsiegel) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseCollection.java:22]: Okay, I understand why they are the way they are, no need to rename.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseRetriever.java
Comment (bsiegel) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseRetriever.java:19]: Leaving as is

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PagedDataCollection.java
Comment (bsiegel) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PagedDataCollection.java:18]: Leaving as-is for this preview.

**Comment by bsiegel** [comment]
Sounds good, can you resolve the remaining comments (just typos and missing headers)? Then we can go ahead and merge.

**Comment by anuchandy** [comment]
Update: This feature is pushed to post preview_1 as per offline discussions within the crew.

**Comment by JianpingChen** [comment]
@bsiegel @vcolin7 @anuchandy I resolved the left over comments. Please take another look and approve if no further concerns.

**Comment by vcolin7** [comment]
@JianpingChen You probably don't have write access to the public repo. I'm not sure who's in charge of assigning such permissions. I will review later today :)

**Comment by weshaggard** [comment]
@JianpingChen @RezaJooyandeh can grant you the permission to be able to merge.

**Comment by RezaJooyandeh** [comment]
@JianpingChen Now do have access :)

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PagedDataCollection.java
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PagedDataCollection.java:34]:
```suggestion
     * Retrieve the first page of the collection
     * @return the first page of the collection
```

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PagedDataCollection.java
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PagedDataCollection.java:42]: nit; We can use the value we just set.
```suggestion
        pages.put(firstPageId, firstPage);
```

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PagedDataCollection.java
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PagedDataCollection.java:52]: What happens if the given `pageId` is `null`? The annotation will not stop users from passing such a value.
```suggestion
        Objects.requireNonNull(pageId);
        P page = pages.get(pageId);
```

File: sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PagedDataCollection.java
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/util/paging/PagedDataCollection.java:67]: nit; we can just use the `pageId` given as an input parameter.
```suggestion
        pages.put(pageId, page);
```

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseCollection.java
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseCollection.java:51]: nit;
```suggestion
        return firstPageResponse;
```

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseCollection.java
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseCollection.java:79]: nit; We can use the `pageId` given as an input parameter.
```suggestion
            pages.put(pageId, pageResponse);
```

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataRetriever.java
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataRetriever.java:26]:
```suggestion
     * Retrieve a page with the given id
     * @param pageId id of the page
     * @param callback callback function to handle the page data along with the response it comes with
```

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java:21]:
```suggestion
 * Represents a collection of pages where page data and the response it comes with it can be obtained asynchronously
 * @param <T> type of the items in the page
 * @param <P> type of the page
```

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java:30]:
```suggestion
     * Constructor requires an asynchronous paged data provider
     * @param asyncPagedDataRetriever an asynchronous paged data provider
```

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java:56]:
```suggestion
     * Retrieve the first page of the collection along with the response it comes with
     * @param callback callback interface for handling the first page and the response it comes with
```

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseCollection.java
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseCollection.java:48]: How likely is it that the `pageId` turns out to be `null`? We don't check for nullness in the other PagedCollection classes.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java:89]:
```suggestion
     * Retrieve a page with the given id
     * @param pageId id of the page
     * @param callback callback interface for handling the page along with the response it comes with
```

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java
Comment (vcolin7) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java:118]: Adding to this for later discussion: We can definitely handle the given `okhttp3.Response` in the callback and map its contents to our own `Response<T>` type. I think we did that in Storage at some point.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/AsyncPagedDataCollection.java:65]: For consistency with `getPage()` below.
```suggestion
                        callback.onSuccess(null, response);
```

**Review by vcolin7 — COMMENTED** [review]
Logic-wise most of the code looks great. I added a couple comments about checking for nullness and a lot of small observations about the Javadoc.

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseCollection.java
Comment (JianpingChen) [reply sdk/core/azure-core/src/main/java/com/azure/android/core/http/responsepaging/PagedDataResponseCollection.java:48]: Added null check wherever missing. Hope this resolves it

**Review by vcolin7 — APPROVED** [review]
Aside from some Javadoc details it looks good to me :)

**Pull request merged**
